tree/20_cfe_basics/time_lib.cf

```diff
@@ -0,0 +1,203 @@
+      "splay_dec" string => execresult("/usr/bin/printf '%d' 0x${splay_hex}", "noshell");
+    test::
+      "splay_dec" string => "${test_splay}";
+
+  classes:
+    any::
+      "test" expression => regcmp("^\d+$", "${test_splay}");
+}
+
+# Generic Scheduler, manage periodicity and splay of execution
+#  The real start time is delayed by a constant number of minutes that is lower than the splay parameter.
+#  This splay time depends on the host properties (IP and hostname)
+#  Everything is calculated in minutes.
+#  This is the basis for other schedulers and is not made to be user friendly.
+#  The outcome are variables that can be used for final decision
+#
+# job_id: a string to identify this job
+# agent_periodicity: how often you run the agent in minutes
+# splay: on how many minutes you want to spread the runs
+# job_periodicity: the time frame between runs in minutes 
+# minutes_from_start: the number of minutes elapsed since the first start of the job (usually: now - any_start_epoch_timestamp/60)
+#
+# If there is an error, a class is raised: job_${job_id}_error
+#    # Sample code to use in the caller 
+#      "job_${job_id}_after_splay" expression => isgreaterthan("${job_schedule_generic.current_run_id}", "${job_schedule_generic.max_run_id}");
+#      "job_${job_id}_after_run" expression => isgreaterthan("${job_schedule_generic.current_run_id}", "${job_schedule_generic.job_run_id}");
+#      "job_${job_id}_run" expression => strcmp("${job_schedule_generic.current_run_id}", "${job_schedule_generic.job_run_id}");
+bundle agent schedule_generic(job_id, agent_periodicity, splay, job_periodicity, minutes_from_start)
+{
+  vars:
+    any::
+      "test" string => "value";
+    noerror::
+      # shift between each run = run_periodicity % agent_periodicity
+      "run_shift_real" string => eval("${job_periodicity} % ${agent_periodicity}", "math", "infix");
+      "run_shift" int => "${run_shift_real}";
+
+      # max_run_id = int(splay / agent periodicity) = min number of agent run in interval
+      "max_run_id_real" string => eval("${splay}/${agent_periodicity}", "math", "infix");
+      "max_run_id" int => "${max_run_id_real}";
+
+      # run_id = int(splay / splay_max * max_run_id) = run id of agent at wich we must run the job
+      "job_run_id_real" string => eval("${splay.splay_dec} * ${max_run_id} / ${splay.splay_max}", "math", "infix");
+      "job_run_id" int => "${job_run_id_real}";
+
+      # elapsed in period = minutes_from_start % run periodicity = duration in minutes since the last start of the run period 
+      "elapsed" string => ifelse("null_job_periodicity", "${minutes_from_start}", eval("${minutes_from_start} % ${job_periodicity}", "math", "infix"));
+
+      # current_run_id = int(elapsed / agent_periodicity) = run id of current run
+      "current_run_id_real" string => eval("${elapsed} / ${agent_periodicity}", "math", "infix");
+      "current_run_id" int => "${current_run_id_real}";
```

Comment:
I don't know how to name them better and their names are explained at the top of the method